### DISCUSSION TITLE
Allow access only to messages, where the message is linked to a project, and the uuid of the user is in the project.members array

### DISCUSSION THREAD

**Discussion by Paul-Browne** [supabase/supabase#31075] [Questions]
So I have 3 tables. Messages, Projects, and Users

Every message belongs to a Project
Every Project has an array of members (project.members)
Every member is an id from the Users table, which links to the auth.users.id

So basically What I need is to allow users who's id is in the project.members array access to all the messages that are in a project, where project.members includes their uuid....

**Comment by GaryAustin1** [comment]
You would likely be better off with a projects_members join table than an array. 

Either way the SB AI or any AI should be able to help with a policy.

**Reply by Paul-Browne** [reply]
So you think having a project_members table would be a better option?

**Reply by GaryAustin1** [reply]
It is faster to add or delete members with a table row than an array. Plus a lot easier to link and managed from the REST api with foreign keys and no ability to add or delete from an array column.

**Reply by Paul-Browne** [reply]
so project_members

id | uuid | project_id 
---|---|---
1 | 123qwe | 4
2 | 123asd | 3
3 | 456qwe | 2
4 | 123zxc | 4
5| 888zzz | 3

**Reply by GaryAustin1** [reply]
Also make user_id ( don’t call it uuid) and project id both foreign keys to their tables and both part of primary key.  Your id column then is not needed. There is a discussion in SB docs database tables section. 

But if array is working for you now depends on your long term needs too switch or not.

**Comment by Paul-Browne** [comment]
SB AI gives me this:

```sql
CREATE POLICY "Allow members of a project to read messages" 
ON messages 
FOR SELECT 
TO authenticated 
USING (
    EXISTS (
        SELECT 1 
        FROM projects 
        WHERE projects.id = messages.project 
        AND auth.uid() = ANY(projects.members)
    )
);
```

which doesn't account for the Users table...

**Reply by GaryAustin1** [reply]
Why would the users table be involved if you have their UUId in the array.

**Reply by Paul-Browne** [reply]
the project.members array is of ids from the Users table.... But yeah, i guess there is no need to query the Users table if I just use the UUID directly in the project.members array 👍
